crates/fuel-core/src/service/genesis/task_manager.rs

```diff
@@ -1,74 +1,126 @@
-use std::future::Future;
+use std::{
+    future::Future,
+    sync::Arc,
+};
 
 use fuel_core_services::StateWatcher;
 use futures::{
+    stream::FuturesUnordered,
     StreamExt,
     TryStreamExt,
 };
 use itertools::Itertools;
 use tokio::task::JoinSet;
+use tokio_util::sync::CancellationToken;
 
 pub struct TaskManager<T> {
     set: JoinSet<anyhow::Result<T>>,
-    cancel: CancellationToken,
+    outside_cancel_listener: MultiCancellationToken,
+    cancel_tasks: CancellationToken,
 }
 
-#[cfg_attr(feature = "test-helpers", derive(Default))]
-#[derive(Clone)]
-pub struct CancellationToken {
-    task_cancellator: tokio_util::sync::CancellationToken,
-    state_watcher: StateWatcher,
+#[async_trait::async_trait]
+pub trait NotifyCancel {
+    async fn wait_until_cancelled(&self) -> anyhow::Result<()>;
+    fn is_cancelled(&self) -> bool;
 }
 
-#[cfg(feature = "test-helpers")]
-impl From<tokio_util::sync::CancellationToken> for CancellationToken {
-    fn from(token: tokio_util::sync::CancellationToken) -> Self {
-        Self {
-            task_cancellator: token,
-            ..Default::default()
+#[async_trait::async_trait]
+impl NotifyCancel for CancellationToken {
+    async fn wait_until_cancelled(&self) -> anyhow::Result<()> {
+        self.cancelled().await;
+        Ok(())
+    }
+    fn is_cancelled(&self) -> bool {
+        self.is_cancelled()
+    }
+}
+
+#[async_trait::async_trait]
+impl NotifyCancel for StateWatcher {
+    async fn wait_until_cancelled(&self) -> anyhow::Result<()> {
+        while !self.is_cancelled() {
+            self.clone().changed().await?;
         }
+
+        Ok(())
     }
+
+    fn is_cancelled(&self) -> bool {
+        let state = self.borrow();
+        state.stopping() || state.stopped()
+    }
+}
+
+/// A token that can be used to monitor multiple cancellation sources.
+#[derive(Default, Clone)]
+pub struct MultiCancellationToken {
+    sources: Vec<Arc<dyn NotifyCancel + Send + Sync>>,
 }
 
-impl CancellationToken {
-    #[cfg(test)]
-    pub async fn cancelled(mut self) -> anyhow::Result<()> {
-        tokio::select! {
-            _ = self.task_cancellator.cancelled() => Ok(()),
-            result = self.state_watcher.wait_stopping_or_stopped() => result
+impl MultiCancellationToken {
+    pub fn from_single(source: impl NotifyCancel + Send + Sync + 'static) -> Self {
+        let mut token = Self::default();
+        token.insert(source);
+        token
+    }
+
+    /// Note: Adding a new source to the token will not affect already running futures.
+    pub fn insert(&mut self, source: impl NotifyCancel + Send + Sync + 'static) {
+        self.sources.push(Arc::new(source));
+    }
+}
+
+#[async_trait::async_trait]
+impl NotifyCancel for MultiCancellationToken {
+    async fn wait_until_cancelled(&self) -> anyhow::Result<()> {
+        if self.sources.is_empty() {
+            futures::future::pending::<()>().await;
+            Ok(())
+        } else {
```

Comment:
It is not clear to me why we want to return to `pending`. More logical looks to return `Ok(())`.

Plus, it seems impossible to have zero sources because you have one constructor that initializes at least one source.